plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String key = keyFromPath(path);
+        if (recursive) {
+            boolean deleted = deletePrefix(path);
+            if (deleted) {
+                deleteObject(key + DIRECTORY_SUFFIX);
+                return true;
+            }
+            else {
+                return deleteObject(key);
             }
         }
-        catch (FileNotFoundException e) {
+        else {
+            if (path.getName().isEmpty()) {
+                return false; // bucket root does not exist
+            }
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(2), true);
+            Iterator<String> objectKeysIterator = Iterators.concat(Iterators.transform(listingsIterator, TrinoS3FileSystem::keysFromListing));
+            if (objectKeysIterator.hasNext()) {
+                String directoryKey = objectKeysIterator.next();
+                if (!Objects.equals(directoryKey, key + PATH_SEPARATOR) || objectKeysIterator.hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                deleteObject(directoryKey);
+            }
+            else {
+                deleteObject(key);
```

Comment:
The previous behavior would have preferred to consider this path to be a "file" and only delete this key if a non-directory object (ie: not marked with content-type: `application/x-directory`) with name `key` existed. When "files" are deleted, we also didn't previously issue `deleteObject(key + DIRECTORY_SUFFIX)` here.

Reply:
I slightly modified the code to always delete the key with `DIRECTORY_SUFFIX` as well.

The option of using `listObjects` for gathering the keys to be deleted makes the code easy to follow. However as we observe, the fact that `MinIO` treats the "directory" object keys and the listing differently compared to AWS S3 makes this option rather wobbly :( .

Reply:
I did this change to cover the MinIO empty directory usecase.
The directory will not get listed in this situation for deletion, and needs to be deleted in the same fashion as an "object". Note that using key corresponding to the provided path (without a trailing `/`) works in MinIO to delete an "empty directory" object.